doc/man/tldc.1
.TH TLDC 1 "December 2021" "libtld 2.x" "User Commands"
.SH NAME
tldc \- Compile TLDs to a binary .tld file
.SH SYNOPSIS
.B tldc
[\fIOPTION\fR]... [\fB--source\fR|\fB-s\fR \fISOURCE\fR] [\fIOUTPUT-FILE\fR]
.SH DESCRIPTION
Compile a set of TLD descriptions found in .ini files in the SOURCE
directory tree to OUTPUT\-FILE which is a binary file format ready
to be used by the tld(3) function. The strings are \fIcompressed\fR
in one super string, and the TLD descriptions are sorted for very
quick binary searches.
.SS "Options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
print this help, then exit
.TP
\fB\-V\fR, \fB\-\-version\fR
print version number, then exit
.TP
\fB\-\-c-file\fR
define the filename of a C file where to output the binary data which can
then be statically compiled in the libtld library; this is used as the
fallback data by the
.BR tld (3)
command
.TP
\fB\-\-output-json\fR
output the data in a JSON file, the filename is the output filename with
the .tld replaced with .json
.TP
\fB\-s\fR, \fB\-\-source\fR \fISOURCE\fR
define the top directory of the directory tree of .ini files to compile
.TP
\fB\-\-verify\fR
run the verification step which makes sure that the saved file can be
reloaded and compared against the existing input data; this step uses the
JSON output data to ensure equality
.PP
The \fB\-\-source\fR command line option changes the source directory tree.
By default this is \fB`/usr/share/libtld/tlds'\fR. That directory includes a
softlink named \fB`extensions'\fR pointing to \fB`/var/lib/libtld/tlds'\fR.
The latter directory is where you are expected to add your own .ini files
if you want to extend the list.
.PP
The \fB\-\-output\-json\fR command line option tells the compile to also
generate a .json file. (\fBTODO:\fR create a version of the
.BR tld (3)
command in JavaScript.)
.PP
The \fB\-\-verify\fR command line option is used at compile time to make
sure that the writing of the .tld file and the later loading of that data
works as expected. You do not need to use that option if you are just
updating your list of TLDs. The input files are verified either way and if
there is an error in one of the .ini file, you will get an error message
with a filename and line number with or without that option.
.PP
The \fB\-\-c\-file\fR command line option is also used at compile time.
Actually, the project creates a temporary tool which sole purpose (nearly)
is to generate the \fBtld_data.c\fR file so the library can include a
statically defined fallback internally. This is a representation of the
binary file in a C array.
.SH AUTHOR
Written by Alexis Wilke <alexis@m2osw.com>.
.SH "REPORTING BUGS"
Report bugs to <https://github.com/m2osw/libtld/issues>.
.br
libtld home page: <https://snapwebsites.org/project/libtld>.
.SH COPYRIGHT
Copyright \(co 2011-2025  Made to Order Software Corp.  All Rights Reserved
.br
License: MIT
.br
This is free software: you are free to change and redistribute it.
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR validate-tld (1),
.BR tld (3),
.BR tld.ini (5).


debian/changelog
libtld (2.0.14.0~noble) noble; urgency=high

  * Updated the TLDs from the suffix file.
  * Fixed the compiler (several issues).
  * Split more brand/entrepreneur files to separate .ini files.
  * Got a rununittests script to work as if I had catch2 tests.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 20 Jul 2025 08:34:13 -0700

libtld (2.0.13.1~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 10 Nov 2023 14:37:26 -0800

libtld (2.0.13.0~jammy) jammy; urgency=high

  * Added missing #include of cstdint for std::uint32_t and others.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 03 Nov 2023 16:28:15 -0700

libtld (2.0.12.0~jammy) jammy; urgency=high

  * Updated TLDs to the latest defined in the public_suffix_list.dat file.
  * Marked many entries as deprecated (not in the public suffix list).
  * Moved some "commercial" domain names to a file under owner's name.
  * Removed some TLDs I added according to WikiPedia and NICs, but if the
    public list does not have them, it's not safe.
  * Various clean ups to the TLD files.
  * Updated the test verifying the public list to go both ways (test present
    and test absent TLDs).
  * Added tld_next_tld() to go through the whole list of TLDs available.
  * Updated test of full URI and internal test to match new domains.
  * Generate an error if the same hunk is found multipletimes in TLD file.
  * Fixed documentation in various places.

 -- Alexis Wilke <alexis@m2osw.com>  Thu, 23 Feb 2023 21:54:40 -0800

libtld (2.0.11.3~bionic) bionic; urgency=high

  * Removed an optimization which doesn't work from the tld comipler.

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 01 Nov 2022 13:23:12 -0700

libtld (2.0.11.2~bionic) bionic; urgency=high

  * This version compiles on my computer.

 -- Alexis Wilke <alexis@m2osw.com>  Mon, 31 Oct 2022 16:54:22 -0700

libtld (2.0.11.1~bionic) bionic; urgency=high

  * The tld_parser is exiting with an exception at the moment. Trying with gdb.

 -- Alexis Wilke <alexis@m2osw.com>  Mon, 31 Oct 2022 16:14:38 -0700

libtld (2.0.11.0~bionic) bionic; urgency=high

  * Removed the debian .dirs files (they were not really used anyway).

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 30 Oct 2022 21:15:08 -0700

libtld (2.0.10.0~bionic) bionic; urgency=high

  * Fixed a couple of error messages about non-quotable characters.

 -- Alexis Wilke <alexis@m2osw.com>  Mon, 04 Jul 2022 11:30:23 -0700

libtld (2.0.9.1~bionic) bionic; urgency=high

  * Updated compat version to v10.

 -- Alexis Wilke <alexis@m2osw.com>  Thu, 19 May 2022 20:33:21 -0700

libtld (2.0.9.0~bionic) bionic; urgency=high

  * Cleaned up the cmake file.

 -- Alexis Wilke <alexis@m2osw.com>  Thu, 19 May 2022 11:20:17 -0700

libtld (2.0.8.1~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 05 Mar 2022 09:52:32 -0800

libtld (2.0.8.0~bionic) bionic; urgency=high

  * Fixed validate-tld, if followed by a port, also remove the port.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 19 Feb 2022 13:59:02 -0800

libtld (2.0.7.0~bionic) bionic; urgency=high

  * Fixed tld_check_uri() so it fails on empty domain names (only TLD is wrong).

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 19 Feb 2022 13:24:51 -0800

libtld (2.0.6.1~bionic) bionic; urgency=high

  * Bumped version to rebuild without the ARM CPU.

 -- Alexis Wilke <alexis@m2osw.com>  Mon, 31 Jan 2022 13:14:10 -0800

libtld (2.0.6.0~bionic) bionic; urgency=high

  * Fixed a script to make use of a variable with the project name.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 30 Jan 2022 20:22:59 -0800

libtld (2.0.5.0~bionic) bionic; urgency=high

  * Applied a fix in a test so it compiles on Hirsute.
  * Reset the length to 0 even if the id is invalid in tld_file_string().

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 04 Jan 2022 20:56:23 -0800

libtld (2.0.4.0~bionic) bionic; urgency=high

  * Looked into fixing the tag merging as one was overflowing.

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 04 Jan 2022 20:28:02 -0800

libtld (2.0.3.0~bionic) bionic; urgency=high

  * Removed the Boost REQUIRED from the main CMakeLists.txt.

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 04 Jan 2022 13:07:39 -0800

libtld (2.0.2.0~bionic) bionic; urgency=high

  * Removed the coverage script. `./mk -c` works as expected.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 02 Jan 2022 07:27:25 -0800

libtld (2.0.1.0~bionic) bionic; urgency=high

  * Further clean ups.
  * Handle memory leaks to prove that we can use the library without such.
  * Fixed the path to the tld_parser tool, we need ./tld_parser.
  * Fixed the add_custom_command() dependencies. tld_parser is not a source.
  * Correctly verify that the tld_file_...() function return non-NULL.
  * Updated the hex tool to output \uXXXX instead of an HTML entity.
  * Removed the lint test for the `tld_data.xml` since the file is gone.
  * Removed the tabs found inside hex.c.
  * Made the path usable in run_all_tests.c so it works running coverage.
  * Made the tests pass even with the sanitizer.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 01 Jan 2022 11:21:03 -0800

libtld (2.0.0.0~bionic) bionic; urgency=high

  * Removed dependency on the Qt library.
  * Implemented our own parser using our own .ini like file format.
  * Fixed all characters with code point higher than 0xFFFF.

 -- Alexis Wilke <alexis@m2osw.com>  Wed, 29 Dec 2021 11:48:51 -0800

libtld (1.6.2.3~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 29 Aug 2021 10:16:10 -0700

libtld (1.6.2.2~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 24 Aug 2021 14:07:21 -0700

libtld (1.6.2.1~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 04 Jun 2021 17:45:11 -0700


tools/tldc.cpp
/* TLD tools -- TLD, domain name, and sub-domain extraction
 * Copyright (c) 2011-2025  Made to Order Software Corp.  All Rights Reserved
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

// libtld
//
#include    <libtld/tld_compiler.h>
#include    <libtld/tld_file.h>


// C++
//
#include    <fstream>
#include    <iostream>
#include    <sstream>


// C
//
#include    <string.h>


class compiler
{
public:
    std::ostream &  error();
    int             exit_code() const;
    void            set_input_path(std::string const & path);
    void            set_output(std::string const & output);
    void            set_c_file(std::string const & c);
    void            set_verify(bool verify);
    void            set_output_json(bool verify);
    void            set_include_offsets(bool include_offsets);
    void            set_verbose(bool verbose);

    void            run();

private:
    void            verify_output(tld_compiler & c);

    int             f_errcnt = 0;
    std::string     f_input_path = std::string();
    std::string     f_output = std::string();
    std::string     f_c_file = std::string();
    bool            f_verify = false;
    bool            f_output_json = false;
    bool            f_include_offsets = false;
    bool            f_verbose = false;
};


std::ostream & compiler::error()
{
    ++f_errcnt;
    return std::cerr;
}


int compiler::exit_code() const
{
    return f_errcnt == 0 ? 0 : 1;
}


void compiler::set_input_path(std::string const & path)
{
    f_input_path = path;
}


void compiler::set_output(std::string const & output)
{
    f_output = output;
}


void compiler::set_c_file(std::string const & c_file)
{
    f_c_file = c_file;
}


void compiler::set_verify(bool verify)
{
    f_verify = verify;
}


void compiler::set_output_json(bool output_json)
{
    f_output_json = output_json;
}


void compiler::set_include_offsets(bool include_offsets)
{
    f_include_offsets = include_offsets;
}


void compiler::set_verbose(bool verbose)
{
    f_verbose = verbose;
}


void compiler::run()
{
    if(f_errcnt != 0)
    {
        // command line found errors, return immediately
        return;
    }

    if(f_input_path.empty())
    {
        ++f_errcnt;
        std::cerr << "error: an input path is required.\n";
        return;
    }

    if(f_output.empty())
    {
        ++f_errcnt;
        std::cerr << "error: an output filename is required.\n";
        return;
    }

    std::cout << "Compiling TLDs from \"" << f_input_path << "\"..." << std::endl;

    tld_compiler c;
    c.set_input_folder(f_input_path);
    c.set_output(f_output);
    c.set_c_file(f_c_file);
    if(!c.compile())
    {
        ++f_errcnt;
        std::cerr
            << "error:"
            << c.get_filename()
            << ":"
            << c.get_line()
            << ": "
            << c.get_errmsg()
            << " (errno: "
            << c.get_errno()
            << ", "
            << strerror(c.get_errno())
            << ")\n";
        return;
    }

    std::cout << "Number of strings:        " << c.get_string_manager().size()         << "\n";
    std::cout << "Longest string:           " << c.get_string_manager().max_length()   << "\n";
    std::cout << "Total string length:      " << c.get_string_manager().total_length() << "\n";
    std::cout << "Included strings:         " << c.get_string_manager().included_count() << " (saved length: " << c.get_string_manager().included_length() << ")\n";
    std::cout << "Mergeable strings:        " << c.get_string_manager().merged_count() << " (saved length: " << c.get_string_manager().merged_length() << ")\n";
    std::cout << "Compressed string length: " << c.get_string_manager().compressed_length() << std::endl;
    // TODO: add info about tags

    if(f_output_json)
    {
        std::string filename;
        std::string::size_type const dot(f_output.rfind('.'));
        if(dot != std::string::npos
        && dot > 0
        && f_output[dot - 1] != '/')
        {
            filename = f_output.substr(0, dot) + ".json";
        }
        else
        {
            filename = f_output + ".json";
        }
        std::ofstream out;
        out.open(filename);
        if(out.is_open())
        {
            c.output_to_json(out, f_include_offsets);
        }
        else
        {
            ++f_errcnt;
            std::cerr
                << "error: could not open JSON output file: \""
                << filename
                << "\".\n";
            return;
        }
    }

    if(f_verify)
    {
        verify_output(c);
    }
}


void compiler::verify_output(tld_compiler & c)
{
    tld_file * file(nullptr);
    auto_free_tld_file auto_free(&file);
    tld_file_error err(tld_file_load(f_output.c_str(), &file));
    if(err != TLD_FILE_ERROR_NONE)
    {
        ++f_errcnt;
        std::cerr << "error: could not load output file \""
            << f_output
            << "\" -- err: "
            << tld_file_errstr(err)
            << " ("
            << static_cast<int>(err)
            << ").\n";
        return;
    }

    // generate a JSON from what was just loaded
    // and it has to match the compiler's JSON
    //
    char * json(tld_file_to_json(file));
    if(json == nullptr)
    {
        ++f_errcnt;
        std::cerr << "error: conversion of file to JSON failed.\n";
        return;
    }
    auto_free_string auto_delete(json);

    // save the verification JSON to a file if we also saved the
    // JSON of the compiler to a file
    //
    if(f_output_json)
    {
        std::string filename;
        std::string::size_type const dot(f_output.rfind('.'));
        if(dot != std::string::npos
        && dot > 0
        && f_output[dot - 1] != '/')
        {
            filename = f_output.substr(0, dot) + "-verify.json";
        }
        else
        {
            filename = f_output + "-verify.json";
        }
        std::ofstream out;
        out.open(filename);
        if(out.is_open())
        {
            out.write(json, strlen(json));
        }
        else
        {
            ++f_errcnt;
            std::cerr
                << "error: could not open JSON output file: \""
                << filename
                << "\".\n";
            return;
        }
    }

    std::stringstream compiler_json;
    c.output_to_json(compiler_json, false);

    if(compiler_json.str() != json)
    {
        ++f_errcnt;
        std::cerr
            << "error: compiler & verification JSON differ."
            << (f_output_json
                ? " Check the two .json output files to see the differences."
                : " Try using the --output-json command line option to get the .json files to find the differences.")
            << "\n";
        return;
    }
}




void usage(char * argv0)
{
    std::string progname(argv0);
    std::string::size_type pos(progname.rfind('/'));
    if(pos != std::string::npos)
    {
        progname = progname.substr(pos + 1);
    }
    std::cout << progname << " v" << LIBTLD_VERSION << "\n";
    std::cout << "Usage: " << progname << " [--opts] [<output>]\n";
    std::cout << "Where --opts is one or more of the following:\n";
    std::cout << "    --help | -h             prints out this help screen and exit\n";
    std::cout << "    --c-file                path and filename to the \"tld_data.c\" file\n";
    std::cout << "    --include-offsets       print offset in comment in .json file\n";
    std::cout << "    --output-json           also save to a .json file\n";
    std::cout << "    --source | -s <folder>  define the source (input) folder\n";
    std::cout << "    --verify                verify loading results and compare against sources\n";
    std::cout << "    --verbose               print out more information about what is happening\n";
    std::cout << "    --version | -V          print out the version and exit\n";
    std::cout << "\n";
    std::cout << "The default source is \"/usr/share/libtld/tlds\".\n";
    std::cout << "The default output is \"/var/lib/libtld/tlds.tld\".\n";
    std::cout << progname << " will not output a C-file or JSON by default.\n";
}


int main(int argc, char * argv[])
{
    compiler tldc;

    for(int i(1); i < argc; ++i)
    {
        if(argv[i][0] == '-')
        {
            if(strcmp(argv[i], "-h") == 0
            || strcmp(argv[i], "--help") == 0)
            {
                usage(argv[0]);
                return 1;
            }
            else if(strcmp(argv[i], "-V") == 0
                 || strcmp(argv[i], "--version") == 0)
            {
                std::cout << LIBTLD_VERSION << std::endl;
                return 1;
            }
            else if(strcmp(argv[i], "-s") == 0
                 || strcmp(argv[i], "--source") == 0)
            {
                ++i;
                if(i >= argc)
                {
                    tldc.error()
                        << "error: argument missing for --source.\n";
                }
                else
                {
                    tldc.set_input_path(argv[i]);
                }
            }
            else if(strcmp(argv[i], "--verify") == 0)
            {
                tldc.set_verify(true);
            }
            else if(strcmp(argv[i], "--c-file") == 0)
            {
                ++i;
                if(i >= argc)
                {
                    tldc.error()
                        << "error: argument missing for --output-c-file.\n";
                }
                else
                {
                    tldc.set_c_file(argv[i]);
                }
            }
            else if(strcmp(argv[i], "--output-json") == 0)
            {
                tldc.set_output_json(true);
            }
            else if(strcmp(argv[i], "--include-offsets") == 0)
            {
                tldc.set_include_offsets(true);
            }
            else if(strcmp(argv[i], "--verbose") == 0)
            {
                tldc.set_verbose(true);
            }
            else
            {
                tldc.error()
                    << "error: unknown command line option \""
                    << argv[i]
                    << "\".\n";
            }
        }
        else
        {
            tldc.set_output(argv[i]);
        }
    }

    tldc.run();

    return tldc.exit_code();
}

// vim: ts=4 sw=4 et
